### PR TITLE
Feature basestats filename

### PR DESCRIPTION
**Pull request by ethanvaughn**
- Modified to use Config.APIName to form the basestats filename rather than suiteName. This allows multiple configs to use the same suite while maintaining separate base stats. eg. API 1x, 2x, and 3x need separate baselines, but require the same suite.

### CODE REVIEW

File: testStrategies/commonToAllTestStrategies.go [hunk 1]
```diff
@@ -410,21 +410,32 @@ func getArrayNameAndIndex(testRunGlobals map[string]interface{}, propPathPart st
...
+		// continue in this case. The user must fix the data issue
+		// before proceeding.
+		if arylen == 0 {
 			log.Errorf("FATAL: Unable to substitute property [%s]: Result array of size 0. Check data criteria for service call.", propertyNameParts[0])
 			os.Exit(1)
 		}
 		// Set the index to a random value.
-		index = randIdx.Intn(len)
+		index = randIdx.Intn(arylen)
 	}
```
Comment (ethanvaughn) [inline testStrategies/commonToAllTestStrategies.go]: Renamed variable to suppress warning about conflicts with built-in "len()" funciton.

File: testStrategies/suiteBasedTesting.go [hunk 2]
```diff
@@ -73,9 +73,9 @@ func executeTestSuite(
 			skipInfrequent = true
 		}
 		// Determine whether "Sparse" items should run this iteration.
-		// [Currently set at 8% (mod 12)]
+		// [Currently set at 3% (mod 30)]
 		skipSparse := false
-		if i%12 != 0 {
+		if i%30 != 0 {
 			skipSparse = true
```
Comment (ethanvaughn) [inline testStrategies/suiteBasedTesting.go]: Found that 8% generated too much load on the sparse calls.

File: report/content.tmpl [hunk 3]
```diff
@@ -61,7 +71,7 @@
                 grid: {
                      x: {
                         lines: [
-                                    {{.JsonTestPartitions}}
+                                    {{.JSONTestPartitions}}
                                 ]
```
Comment (ethanvaughn) [inline report/content.tmpl]: golint

File: report/content.tmpl [hunk 4]
```diff
@@ -136,12 +146,22 @@
...
         <script>
            var barChartJS = c3.generate({
+               size: {
+                    height: 500
+                },
                 data: {
                     columns: [
-                        {{.JsonTimeArray}}
+                        {{.JSONTimeArray}}
                     ],
```
Comment (ethanvaughn) [inline report/content.tmpl]: golint

File: report/content.tmpl [hunk 5]
```diff
@@ -136,12 +146,22 @@
...
                     type: 'bar'
                 },
+                legend: {
+                    show: true,
+                    position: 'inset',
+                    inset: {
+                        anchor: 'top-right'
+                    }
+                },
                 zoom: {
```
Comment (ethanvaughn) [inline report/content.tmpl]: Move the legend to be out of the clutter.

File: report/content.tmpl [hunk 6]
```diff
@@ -151,12 +171,17 @@
                     }
                 },
                 axis: {
-                y: {
-                    label: 'Avg Resp Time (MilliSeconds)'
-                },
+                    y: {
+                        label: 'Avg Resp Time (MilliSeconds)'
+                    },
                     x: {
```
Comment (ethanvaughn) [inline report/content.tmpl]: Fixed indent issues.

File: report/content.tmpl [hunk 7]
```diff
@@ -151,12 +171,17 @@
...
                         type: 'category',
-                        categories: {{.JsonTimeServiceNames}}
+                        categories: {{.JSONTimeServiceNames}},
+                        tick: {
+                            rotate: 90,
+                            multiline: false
+                        },
+                        height: 200
                     }
```
Comment (ethanvaughn) [inline report/content.tmpl]: golint

File: perfTestUtils/ReportGenerator.go [hunk 8]
```diff
@@ -8,13 +8,14 @@ import (
...
+	"sort"
 )
 
 type perfStatsModel struct {
 	BasePerfStats        *BasePerfStats
 	PerfStats            *PerfStats
 	Config               *Config
-	JsonTimeServiceNames template.JS
+	JSONTimeServiceNames template.JS
 	TestStrategy         string
```
Comment (ethanvaughn) [inline perfTestUtils/ReportGenerator.go]: golint

File: perfTestUtils/ReportGenerator.go [hunk 9]
```diff
@@ -33,7 +34,7 @@ func (p *perfStatsModel) IsServiceTimePass(s string) bool {
 }
 
 func (p *perfStatsModel) IsTimePass() bool {
-	for k, _ := range p.BasePerfStats.BaseServiceResponseTimes {
+	for k := range p.BasePerfStats.BaseServiceResponseTimes {
 		if !p.IsServiceTimePass(k) {
```
Comment (ethanvaughn) [inline perfTestUtils/ReportGenerator.go]: golint

File: perfTestUtils/ReportGenerator.go [hunk 10]
```diff
@@ -45,19 +46,19 @@ func (p *perfStatsModel) PeakMemoryVariancePercentage() float64 {
 	return float64(CalcPeakMemoryVariancePercentage(p.BasePerfStats.BasePeakMemory, p.PerfStats.PeakMemory))
 }
 
-func MemoryMB(pm uint64) float64 {
+func memoryMB(pm uint64) float64 {
 	return float64((float32(pm) / float32(1024)) / float32(1024))
```
Comment (ethanvaughn) [inline perfTestUtils/ReportGenerator.go]: golint (this is not an exported function)

File: perfTestUtils/ReportGenerator.go [hunk 11]
```diff
@@ -45,19 +46,19 @@ func (p *perfStatsModel) PeakMemoryVariancePercentage() float64 {
...
 }
 
-func Div(num int64, den int64) float64 {
+func div(num int64, den int64) float64 {
 	return float64(float32(num) / float32(den))
 }
 
-func JsonMemoryArray(name string, array []uint64) template.JS {
+func jsonMemoryArray(name string, array []uint64) template.JS {
 	jsonMemoryAudit := []byte("['" + name + "',")
```
Comment (ethanvaughn) [inline perfTestUtils/ReportGenerator.go]: golint (not exported)

File: perfTestUtils/ReportGenerator.go [hunk 12]
```diff
@@ -67,19 +68,35 @@ func JsonMemoryArray(name string, array []uint64) template.JS {
 	return template.JS(jsonMemoryAudit)
 }
 
-func (p *perfStatsModel) JsonTestPartitions() template.JS {
+func (p *perfStatsModel) JSONTestPartitions() template.JS {
 	testpatritions := []byte("")
```
Comment (ethanvaughn) [inline perfTestUtils/ReportGenerator.go]: golint

File: perfTestUtils/ReportGenerator.go [hunk 13]
```diff
@@ -67,19 +68,35 @@ func JsonMemoryArray(name string, array []uint64) template.JS {
...
+	for k := range p.PerfStats.ServiceResponseTimes {
+		sortedkeys[i] = k
+		i++
+	}
+	sort.Strings(sortedkeys)
+
+	// Format the chart data.
+	for _, name := range sortedkeys {
+		averageServiceResponseTime := p.PerfStats.ServiceResponseTimes[name]
 		baseTimeMillis := float64(float32(p.BasePerfStats.BaseServiceResponseTimes[name]) / float32(1000000))
```
Comment (ethanvaughn) [inline perfTestUtils/ReportGenerator.go]: Modified to sort the service names on the chart to match the table.

File: perfTestUtils/ReportGenerator.go [hunk 14]
```diff
@@ -100,12 +117,14 @@ func (p *perfStatsModel) JsonTimeArray() template.JS {
 	serviceResponseTimesBase = append(serviceResponseTimesBase, []byte("],")...)
 	serviceResponseTimesTest = append(serviceResponseTimesTest, []byte("]")...)
 	serviceNames = append(serviceNames, []byte("']")...)
-	p.JsonTimeServiceNames = template.JS(serviceNames)
+	p.JSONTimeServiceNames = template.JS(serviceNames)
```
Comment (ethanvaughn) [inline perfTestUtils/ReportGenerator.go]: golint

File: perfTestUtils/ReportGenerator.go [hunk 15]
```diff
@@ -100,12 +117,14 @@ func (p *perfStatsModel) JsonTimeArray() template.JS {
...
 
 	serviceResponseTimesBase = append(serviceResponseTimesBase, serviceResponseTimesTest...)
 	return template.JS(serviceResponseTimesBase)
 }
 
+// GenerateTemplateReport wraps the generateTemplate() function that creates
+// the final performance report html. 
 func GenerateTemplateReport(basePerfstats *BasePerfStats, perfStats *PerfStats, configurationSettings *Config, fs FileSystem, testSuiteName string, testStrategy string) {
```
Comment (ethanvaughn) [inline perfTestUtils/ReportGenerator.go]: golint

File: perfTestUtils/ReportGenerator.go [hunk 16]
```diff
@@ -156,7 +175,7 @@ func generateTemplate(bstats *BasePerfStats, pstats *PerfStats, configurationSet
 			}
 			_, err = s1.New(tname).Parse(string(header))
 			if err != nil {
-				return fmt.Errorf("Error parsing template %v: %v\n", tname, err)
+				return fmt.Errorf("Error parsing template %v: %v", tname, err)
 			}
```
Comment (ethanvaughn) [inline perfTestUtils/ReportGenerator.go]: golint

File: report/content.tmpl [hunk 17]
```diff
@@ -47,6 +47,16 @@
...
+                    height: 500
+                },
+                legend: {
+                    show: true,
+                    position: 'inset',
+                    inset: {
+                        anchor: 'top-right'
+                    }
+                },
                 zoom: {
```
Comment (ethanvaughn) [inline report/content.tmpl]: move the legend to the top of the chart to be out of the clutter

File: report/content.tmpl [hunk 18]
```diff
@@ -151,12 +171,17 @@
...
-                    label: 'Avg Resp Time (MilliSeconds)'
-                },
+                    y: {
+                        label: 'Avg Resp Time (MilliSeconds)'
+                    },
                     x: {
                         type: 'category',
-                        categories: {{.JsonTimeServiceNames}}
+                        categories: {{.JSONTimeServiceNames}},
+                        tick: {
```
Comment (ethanvaughn) [inline report/content.tmpl]: golint

File: report/content.tmpl (hunk 7 above)
Comment (ethanvaughn) [inline report/content.tmpl]: rotate the service name labels to be readable

File: report/content.tmpl [hunk 19]
```diff
@@ -151,12 +171,17 @@
...
-                y: {
-                    label: 'Avg Resp Time (MilliSeconds)'
-                },
+                    y: {
+                        label: 'Avg Resp Time (MilliSeconds)'
+                    },
                     x: {
                         type: 'category',
-                        categories: {{.JsonTimeServiceNames}}
+                        categories: {{.JSONTimeServiceNames}},
```
Comment (ethanvaughn) [inline report/content.tmpl]: golint

**Comment by flynnmichaelj** [comment]
Approved

**Pull request merged**